pkg/bmt/bmt_test.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package bmt_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/bmt"
+	"github.com/ethersphere/bee/pkg/bmt/reference"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// testPoolSize is the number of bmt trees the pool keeps when
+	testPoolSize = 16
+
+	// the actual data length generated (could be longer than max datalength of the BMT)
+	BufferSize = 4128
+
+	// segmentCount is the maximum number of segments of the underlying chunk
+	// Should be equal to max-chunk-data-size / hash-size
+	// Currently set to 128 == 4096 (default chunk size) / 32 (sha3.keccak256 size)
+	testSegmentCount = 128
+)
+
+var (
+	testSegmentCounts = []int{1, 2, 3, 4, 5, 8, 9, 15, 16, 17, 32, 37, 42, 53, 63, 64, 65, 111, 127, 128}
+	hashSize          = swarm.NewHasher().Size()
+	testData          = make([]byte, BufferSize)
+
+	seed = time.Now().Unix()
+)
+
+func refHash(count, n int, data []byte) ([]byte, error) {
+	rbmt := reference.NewRefHasher(swarm.NewHasher(), count)
+	refNoMetaHash, err := rbmt.Hash(data[:n])
+	if err != nil {
+		return nil, err
+	}
+	return bmt.Sha3hash(bmt.LengthToSpan(int64(n)), refNoMetaHash)
+}
+
+// Hash hashes the data and the span using the bmt hasher
+func syncHash(h *bmt.Hasher, n int, data []byte) ([]byte, error) {
+	h.Reset()
+	h.SetHeaderInt64(int64(n))
+	_, err := h.Write(data[:n])
```

Comment:
Theoretically you're right but when you have a look at how this is used across tests.... this starts to get really ugly. Especially the ones that actually use an uninitialized testData slice :man_shrugging: I tried to do the fix but backed away very quickly since it was getting out of hand. @zelig care to make this pretty? or should we do this at a later PR?